include/measurement_kit/nettest.hpp

```diff
@@ -0,0 +1,1922 @@
+    void on_status_started(const events::StatusStartedEvent &event) override;
+
+    /// Logs the "status.update.performance" event on std::clog.
+    void on_status_update_performance(const events::StatusUpdatePerformanceEvent &event) override;
+
+    /// Logs the "status.update.websites" event on std::clog.
+    void on_status_update_websites(const events::StatusUpdateWebsitesEvent &event) override;
+
+    /// Logs the "task_terminated" event on std::clog.
+    void on_task_terminated(const events::TaskTerminatedEvent &event) override;
+
+    ~NoisyRouter() noexcept override;
+};
+
+} // namespace routers
+
+/// Manages the lifecycle of a nettest. This API mirrors as closely as
+/// possible to FFI API provided by <measurement_kit/ffi.h>.
+class Nettest {
+  public:
+    // C++ object model
+    // ````````````````
+
+    /// Creates a nettest instance. This will not start the actual test. You
+    /// need to use one of the start_XXX() methods to do that.
+    Nettest() noexcept;
+
+    /// Explicitly deleted copy constructor.
+    Nettest(const Nettest &) noexcept = delete;
+
+    /// Explicitly deleted copy assignment.
+    Nettest &operator=(const Nettest &) noexcept = delete;
+
+    /// Explicitly deleted move constructor.
+    Nettest(Nettest &&) noexcept = delete;
+
+    /// Explicitly deleted move assignment.
+    Nettest &operator=(Nettest &&) noexcept = delete;
+
+    /// Wait for nettest to terminate and destroy resources.
+    ~Nettest() noexcept;
+
+    // Starting a nettest
+    // ``````````````````
+
+    /// Starts a "CaptivePortal" nettest.
+    /// \param settings contains the settings configuring this nettest.
+    /// \param router is the router used to route nettest events. If the
+    /// router is `nullptr`, we'll use an internal dummy router.
+    /// \return true on success, false on failure. There are two reasons why
+    /// this method could fail. Either the provided settings are not JSON
+    /// serializable because there was a string that was not valid UTF-8, or
+    /// you already used this class to start a nettest. In both cases, log
+    /// messages will be routed, so you will know what happened.
+    bool start_captive_portal(settings::CaptivePortalSettings settings, routers::Router *router) noexcept;
+
+    /// Starts a "Dash" nettest.
+    /// \param settings contains the settings configuring this nettest.
+    /// \param router is the router used to route nettest events. If the
+    /// router is `nullptr`, we'll use an internal dummy router.
+    /// \return true on success, false on failure. There are two reasons why
+    /// this method could fail. Either the provided settings are not JSON
+    /// serializable because there was a string that was not valid UTF-8, or
+    /// you already used this class to start a nettest. In both cases, log
+    /// messages will be routed, so you will know what happened.
+    bool start_dash(settings::DashSettings settings, routers::Router *router) noexcept;
```

Comment:
Also, `router` is an argument so that `log` errors caused by incorrect usage (e.g. starting more than one test with this class) are not only apparent by the `bool` return value but also by explicit log messages.